File: WeatherSense/trunk/VantagePro2Driver/src/vantagepro2/LoopPacket.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Diagnostics;

namespace VantagePro2 {
    public class LoopPacket {
        public enum BaroTrendEnum {
            FALLING_RAPIDLY = 196,
            FALLING_SLOWLY = 236,
            STEADY = 0,
            RISING_SLOWLY = 20,
            RISING_RAPIDLY = 60
        }
        public enum Forecast {
            SUNNY = 8,
            PARTLY_CLOUDY = 6,
            MOSTLY_CLOUDY = 2,
            MOSTLY_CLOUDY_WITH_RAIN = 3,
            MOSTLY_CLOUDY_WITH_SNOW = 18,
            MOSTLY_CLOUDY_WITH_RAIN_OR_SNOW = 19,
            PARTLY_CLOUDY_WITH_RAIN_LATER = 7,
            PARTLY_CLOUDY_WITH_SNOW_LATER = 22,
            PARTLY_CLOUDY_WITH_RAIN_OR_SNOW_LATER = 23
        }

        private static readonly int TEMPERATURE_OFFSET = 90;
        public static readonly int NUM_EXTRA_TEMPERATURES = 7;
        public static readonly int NUM_EXTRA_HUMIDITIES = 7;
        public static readonly int NUM_SOIL_TEMPERATURES = 4;
        public static readonly int NUM_LEAF_TEMPERATURES = 3;
        public static readonly int NUM_SOIL_MOISTURES = 4;
        public static readonly int NUM_LEAF_WETNESSES = 3;
        private static readonly float TEMPERATURE_SCALE = 10.0f;
        private static readonly int INVALID_EXTRA_TEMPERATURE = 255;
        private static readonly int INVALID_EXTRA_HUMIDITY = 255;
        private static readonly int INVALID_UV_INDEX = 255;
        private static readonly int INVALID_LEAF_WETNESS = 255;
        private static readonly int INVALID_SOIL_MOISTURE = 255;
        private static readonly int INVALID_SOLAR_RADIATION = 32767;


        private int nextRecord;
        public int NextRecord {
            get { return nextRecord; }
            set { nextRecord = value; }
        }

        private BaroTrendEnum baroTrend;
        public BaroTrendEnum BaroTrend {
            get { return baroTrend; }
        }

	private float outsideTemp;
        public float OutsideTemp {
            get { return outsideTemp; }
        }
        
	private float insideTemp;
        public float InsideTemp {
            get { return insideTemp; }
        }

	private int outsideHum;
        public int OutsideHumidity {
            get { return outsideHum; }
        }

        private int insideHum;
        public int InsideHumidity {
            get { return insideHum; }
        }

        private float windSpeed;
        public float WindSpeed {
            get { return windSpeed; }
        }
        private float avgWindSpeed10Min;
        public float AvgWindSpeed10Min {
            get { return avgWindSpeed10Min; }
        }

        private int windDirection;
        public int WindDirection {
            get { return windDirection; }
        }

        private float barometer;
        public float Barometer {
            get { return barometer; }
        }

        private float rainRate;
        public float RainRate {
            get { return rainRate; }
        }

        private float stormRain;
        public float StormRain {
            get { return stormRain; }
        }

        private DateTime stormStart;
        public DateTime StormStart {
            get { return stormStart; }
        }

        private float dayRain;
        public float DayRain {
            get { return dayRain; }
        }

        private float monthRain;
        public float MonthRain {
            get { return monthRain; }
        }

        private float yearRain;
        public float YearRain {
            get { return yearRain; }
        }

        private int uvIndex;
        public int UvIndex {
            get { return uvIndex; }
            set { uvIndex = value; }
        }
        public bool isUvIndexValid() {
            return uvIndex != INVALID_UV_INDEX;
        }

        private int solarRadiation;
        public int SolarRadiation {
            get { return solarRadiation; }
            set { solarRadiation = value; }
        }

        public bool isSolarRadiationValid() {
            return solarRadiation != INVALID_SOLAR_RADIATION;
        }

        private float dayET;
        public float DayET {
            get { return dayET; }
            set { dayET = value; }
        }

        private float monthET;
        public float MonthET {
            get { return monthET; }
            set { monthET = value; }
        }

        private float yearET;
        public float YearET {
            get { return yearET; }
            set { yearET = value; }
        }

	private int[] soilMoist = new int[NUM_SOIL_MOISTURES];
        private int[] leafWetness = new int[NUM_LEAF_WETNESSES];
	private float[] leafTemp = new float[NUM_LEAF_TEMPERATURES];
        private bool[] leafTempValid = new bool[NUM_LEAF_TEMPERATURES];
        private float[] soilTemp = new float[NUM_SOIL_TEMPERATURES];
        private bool[] soilTempValid = new bool[NUM_SOIL_TEMPERATURES];

        private float[] tempExtra = new float[NUM_EXTRA_TEMPERATURES];
        private bool[] tempExtraValid = new bool[NUM_EXTRA_TEMPERATURES];
       
        private int[] humidityExtra = new int[NUM_EXTRA_HUMIDITIES];

        private int transmitterBatteryStatus;

        private float consoleBatteryVoltage;
        public float ConsoleBatteryVoltage {
            get { return consoleBatteryVoltage; }
        }

        private Forecast forecastIcon;
        public Forecast ForecastIcon {
            get { return forecastIcon; }
        }

        private int forecastRule;
        public int ForecastRule {
            get { return forecastRule; }
        }

	private DateTime sunRiseTime;
	private DateTime sunSetTime;

        public LoopPacket() {
        }

        public bool isTransmitterBatteryGood(int index) {
            return (transmitterBatteryStatus & (1 << (index - 1))) == 0;
        }

        public bool isExtraTemperatureValid(int index) {
            return tempExtraValid[index];
        }

        public float getExtraTemperature(int index) {
            return tempExtra[index];
        }

        public bool isExtraHumidityValid(int index) {
            return humidityExtra[index] != INVALID_EXTRA_HUMIDITY;
        }

        public int getExtraHumidity(int index) {
            return humidityExtra[index];
        }

        public int getLeafWetness(int index) {
            return leafWetness[index];
        }

        public bool isLeafWetnessValid(int index) {
            return leafWetness[index] != INVALID_LEAF_WETNESS;
        }

        public int getSoilMoisture(int index) {
            return soilMoist[index];
        }

        public bool isSoilMoistureValid(int index) {
            return soilMoist[index] != INVALID_SOIL_MOISTURE;
        }

        public float getSoilTemperature(int index) {
            return soilTemp[index];
        }

        public bool isSoilTemperatureValid(int index) {
            return soilTempValid[index];
        }

        public float getLeafTemperature(int index) {
            return leafTemp[index];
        }

        public bool isLeafTemperatureValid(int index) {
            return leafTempValid[index];
        }

        //
        // A storm is occuring if the storm start year is not 1. In C# a DateTime is initialized to
        // 1/1/1. So if the storm start time in the LOOP packet is -1, then the DateTime will be the
        // default value. In addition, the last couple of LOOP packets that have a valid storm start will
        // report a storm rain total of 0.0 inches. This may be an indicator that the storm has stopped,
        // but we are not using that at this point in time. By definition the storm rain has to be > 0, so
        // we will stop reporting an ongoing storm if the storm rain is 0.0
        //
        public bool isStormOngoing() {
            return stormStart.Year != 1 && stormRain > 0.0;
        }

        public bool parseLoopPacket(byte[] buffer, float rainInterval) {
            if ((char)buffer[0] != 'L' || (char)buffer[1] != 'O' || (char)buffer[2] != 'O')
                return false;

            if ((char)buffer[3] != 'P') {
                switch (buffer[3]) {
                    case 196:
                        baroTrend = BaroTrendEnum.FALLING_RAPIDLY;
                        break;
                    case 236:
                        baroTrend = BaroTrendEnum.FALLING_SLOWLY;
                        break;
                    case 0:
                        baroTrend = BaroTrendEnum.STEADY;
                        break;
                    case 20:
                        baroTrend = BaroTrendEnum.RISING_SLOWLY;
                        break;
                    case 60:
                        baroTrend = BaroTrendEnum.RISING_RAPIDLY;
                        break;
                    default:
                        return false;
                }
            }

            int packetType = (int)buffer[4];

            if (packetType != 0) {
                Debug.Print("Invalid packet type for LOOP packet: {0}", packetType);
                return false;
            }

            nextRecord = BitConverter.ToInt16(buffer,5);
            barometer = UnitConverter.toInchesHg((float)BitConverter.ToInt16(buffer, 7) / 1000.0F);
            insideTemp = UnitConverter.toCelcius((float)BitConverter.ToInt16(buffer, 9) / TEMPERATURE_SCALE);
            insideHum = buffer[11];
            outsideTemp = UnitConverter.toCelcius(BitConverter.ToInt16(buffer, 12) / TEMPERATURE_SCALE);
            windSpeed = UnitConverter.toMetersPerSecond((float)buffer[14]);
            avgWindSpeed10Min = UnitConverter.toMetersPerSecond((float)buffer[15]);
            windDirection = BitConverter.ToInt16(buffer, 16);
            //
            // Zero degrees from the weather station means no wind, translate 360 to 0 for north
            //
            if (windDirection == 360)
                windDirection = 0;

            for (int i = 0; i < NUM_EXTRA_TEMPERATURES; i++) {
                int temperature = buffer[18 + i];
                tempExtraValid[i] = temperature != INVALID_EXTRA_TEMPERATURE;
                tempExtra[i] = UnitConverter.toCelcius(temperature - TEMPERATURE_OFFSET);
            }

            for (int i = 0; i < NUM_SOIL_TEMPERATURES; i++) {
                int temperature = buffer[25 + i];
                soilTempValid[i] = temperature != INVALID_EXTRA_TEMPERATURE;
                soilTemp[i] = UnitConverter.toCelcius((float)temperature - TEMPERATURE_OFFSET);
            }

            for (int i = 0; i < NUM_LEAF_TEMPERATURES; i++) {
                int temperature = buffer[29 + i];
                leafTempValid[i] = temperature != INVALID_EXTRA_TEMPERATURE;
                leafTemp[i] = UnitConverter.toCelcius(temperature - TEMPERATURE_OFFSET);
            }

            outsideHum = buffer[33];

            for (int i = 0; i < NUM_EXTRA_HUMIDITIES; i++) {
                humidityExtra[i] = buffer[34 + i];
            }

            rainRate = UnitConverter.toMillimeter((float)BitConverter.ToInt16(buffer, 41) / 100.0F);

            uvIndex = buffer[43];
            solarRadiation = BitConverter.ToInt16(buffer, 44);

            stormRain = UnitConverter.toMillimeter((float)BitConverter.ToInt16(buffer, 46) / 100.0F);

            short time = BitConverter.ToInt16(buffer, 48);

            if (time != -1)
                stormStart = extractDate(time);

            dayRain = UnitConverter.toMillimeter((float)BitConverter.ToInt16(buffer, 50) * rainInterval);
            monthRain = UnitConverter.toMillimeter((float)BitConverter.ToInt16(buffer, 52) * rainInterval);
            yearRain = UnitConverter.toMillimeter((float)BitConverter.ToInt16(buffer, 54) * rainInterval);

            int idayET = BitConverter.ToInt16(buffer, 56);
            int imonthET = BitConverter.ToInt16(buffer, 58);
            int iyearET = BitConverter.ToInt16(buffer, 60);
            Debug.Print("ET: Day={0}, Month={1}, Year={2}", idayET, imonthET, iyearET);
            dayET = UnitConverter.toMillimeter((float)BitConverter.ToInt16(buffer, 56) / 1000F);
            monthET = UnitConverter.toMillimeter((float)BitConverter.ToInt16(buffer, 58) / 100F);
            yearET = UnitConverter.toMillimeter((float)BitConverter.ToInt16(buffer, 60) / 100F);

            for (int i = 0; i < NUM_SOIL_MOISTURES; i++)
                soilMoist[i] = buffer[62 + i];

            for (int i = 0; i < NUM_LEAF_WETNESSES; i++)
                leafWetness[i] = buffer[66 + i];

            transmitterBatteryStatus = (int)buffer[86];
            Debug.Print("Transmitter Battery Status: {0}", transmitterBatteryStatus);
            consoleBatteryVoltage = (((int)BitConverter.ToInt16(buffer, 87) * 300) / 512) / 100.0F;
            forecastIcon = (Forecast)buffer[89];
            forecastRule = buffer[90];

            DateTime today = DateTime.Today;
            int sunrise = BitConverter.ToInt16(buffer, 91);
            int sunset = BitConverter.ToInt16(buffer, 93);

            sunRiseTime = new DateTime(today.Year, today.Month, today.Day, sunrise / 100, sunrise % 100, 0);
            sunSetTime = new DateTime(today.Year, today.Month, today.Day, sunset / 100, sunset % 100, 0);
        
            if ((char)buffer[95] != CommandStrings.LINE_FEED || (char)buffer[96] != CommandStrings.CARRIAGE_RETURN) {
                Console.WriteLine("<LF><CR> not found");
                return false;
            }

            //Console.WriteLine(this.ToString());

            return VantagePro2Station.checkCRC(buffer, 97);
        }

        public static DateTime extractDate(short time) {
            int year = (time & 0x3F) + 2000;
            int day = (time >> 7) & 0x1F;
            int month = (time >> 12) & 0xF;

            return new DateTime(year, month, day);
        }

        public override String ToString() {
            StringBuilder sb = new StringBuilder();
            sb.AppendFormat("Next Record: {0}\n", nextRecord);
            sb.AppendFormat("Barometer: {0}  Trend: {1}\n", barometer, baroTrend);
            sb.AppendFormat("Outdoor Temperature: {0}  Outdoor Humidity: {1}\n", outsideTemp, outsideHum);
            sb.AppendFormat("Indoor Temperature: {0}  Indoor Humidity: {1}\n", insideTemp, insideHum);
            sb.AppendFormat("Extra Temperatures:");
            for (int i = 0; i < NUM_EXTRA_TEMPERATURES; i++)
                sb.AppendFormat(" [{0}]: {1}", i, tempExtra[i]);
            sb.Append("\n");
            sb.AppendFormat("Extra Humidities:");
            for (int i = 0; i < NUM_EXTRA_HUMIDITIES; i++)
                sb.AppendFormat(" [{0}]: {1}", i, humidityExtra[i]);
            sb.Append("\n");
            sb.AppendFormat("Wind: {0} MPH from {1}  10 Minute Avg Speed: {2}\n", windSpeed, windDirection, avgWindSpeed10Min);
            sb.AppendFormat("Rain: Rate: {0}  Storm: {1}  Day: {2}  Month: {3}  Year: {4}\n", rainRate, stormRain, dayRain, monthRain, yearRain);
            sb.AppendFormat("Leaf Wetnesses:");
            for (int i = 0; i < NUM_LEAF_WETNESSES; i++)
                sb.AppendFormat(" [{0}]: {1}", i, leafWetness[i]);
            sb.Append("\n");
            sb.AppendFormat("Soil Moistures:");
            for (int i = 0; i < NUM_SOIL_MOISTURES; i++)
                sb.AppendFormat(" [{0}]: {1}", i, soilMoist[i]);
            sb.Append("\n");
            sb.AppendFormat("Soil Temperatures:");
            for (int i = 0; i < NUM_SOIL_TEMPERATURES; i++)
                sb.AppendFormat(" [{0}]: {1}", i, soilTemp[i]);
            sb.Append("\n");
            sb.AppendFormat("Leaf Temperatures:");
            for (int i = 0; i < NUM_LEAF_TEMPERATURES; i++)
                sb.AppendFormat("  [{0}]: {1}", i, leafTemp[i]);
            sb.Append("\n");
            sb.AppendFormat("Transmitter Battery Status: 0x{0:x}\n", transmitterBatteryStatus);
            sb.AppendFormat("Console Battery Voltage: {0}\n", consoleBatteryVoltage);
            sb.AppendFormat("Forecast: {0}\n", forecastIcon);
            sb.AppendFormat("Forecast Rule: {0}\n", forecastRule);
            sb.AppendFormat("Sunrise: {0}  Sunset: {1}", sunRiseTime, sunSetTime);
            sb.Append("=========================\n");

            return sb.ToString();
        }
    }
}

File: VantagePro2Driver/VP2++/src/VP2/LoopPacket.cpp
/* 
 * Copyright (C) 2022 Bruce Beisel
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <time.h>
#include <iostream>
#include "BitConverter.h"
#include "UnitConverter.h"
#include "VP2Constants.h"
#include "VP2Decoder.h"
#include "VantagePro2CRC.h"
#include "LoopPacket.h"

using namespace std;

namespace vp2 {

static const char *ALARM_STRINGS[][8] = {
    {
        "Falling Barometer Trend",
        "Rising Barometer Trend",
        "Low Indoor Temperature",
        "High Indoor Temperature",
        "Low Indoor Humidity",
        "High Indoor Humidity",
        "Time",
        nullptr
    },
    {
        "High Rain Rate",
        "15 Minute Rain",
        "24 Hour Rain",
        "Storm Total Rain",
        "Daily ET",
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Low Outdoor Temperature",
        "High Outdoor Temperature",
        "Wind Speed",
        "10 Minute Average Wind Speed",
        "Low Dew Point",
        "High Dew Point",
        "High Heat Index",
        "Low Wind Chill"
    },
    {
        "High THSW",
        "High Solar Radiation",
        "High UV",
        "UV Dose",
        "UV Dose Manually Cleared",
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Low Outdoor Humidity",
        "High Outdoor Humidity",
        nullptr,
        nullptr,
        nullptr,
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Extra Temperature/Humidity 1 - Low Temperature",
        "Extra Temperature/Humidity 1 - High Temperature",
        "Extra Temperature/Humidity 1 - Low Humidity",
        "Extra Temperature/Humidity 1 - High Humidity",
        nullptr,
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Extra Temperature/Humidity 2 - Low Temperature",
        "Extra Temperature/Humidity 2 - High Temperature",
        "Extra Temperature/Humidity 2 - Low Humidity",
        "Extra Temperature/Humidity 2 - High Humidity",
        nullptr,
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Extra Temperature/Humidity 3 - Low Temperature",
        "Extra Temperature/Humidity 3 - High Temperature",
        "Extra Temperature/Humidity 3 - Low Humidity",
        "Extra Temperature/Humidity 3 - High Humidity",
        nullptr,
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Extra Temperature/Humidity 4 - Low Temperature",
        "Extra Temperature/Humidity 4 - High Temperature",
        "Extra Temperature/Humidity 4 - Low Humidity",
        "Extra Temperature/Humidity 4 - High Humidity",
        nullptr,
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Extra Temperature/Humidity 5 - Low Temperature",
        "Extra Temperature/Humidity 5 - High Temperature",
        "Extra Temperature/Humidity 5 - Low Humidity",
        "Extra Temperature/Humidity 5 - High Humidity",
        nullptr,
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Extra Temperature/Humidity 6 - Low Temperature",
        "Extra Temperature/Humidity 6 - High Temperature",
        "Extra Temperature/Humidity 6 - Low Humidity",
        "Extra Temperature/Humidity 6 - High Humidity",
        nullptr,
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Extra Temperature/Humidity 7 - Low Temperature",
        "Extra Temperature/Humidity 7 - High Temperature",
        "Extra Temperature/Humidity 7 - Low Humidity",
        "Extra Temperature/Humidity 7 - High Humidity",
        nullptr,
        nullptr,
        nullptr,
        nullptr
    },
    {
        "Soil/Leaf 1 - Low Leaf Wetness",
        "Soil/Leaf 1 - High Leaf Wetness",
        "Soil/Leaf 1 - Low Soil Moisture",
        "Soil/Leaf 1 - High Soil Moisture",
        "Soil/Leaf 1 - Low Leaf Temperature",
        "Soil/Leaf 1 - High Leaf Temperature",
        "Soil/Leaf 1 - Low Soil Temperature",
        "Soil/Leaf 1 - High Soil Temperature",
    },
    {
        "Soil/Leaf 2 - Low Leaf Wetness",
        "Soil/Leaf 2 - High Leaf Wetness",
        "Soil/Leaf 2 - Low Soil Moisture",
        "Soil/Leaf 2 - High Soil Moisture",
        "Soil/Leaf 2 - Low Leaf Temperature",
        "Soil/Leaf 2 - High Leaf Temperature",
        "Soil/Leaf 2 - Low Soil Temperature",
        "Soil/Leaf 2 - High Soil Temperature",
    },
    {
        "Soil/Leaf 3 - Low Leaf Wetness",
        "Soil/Leaf 3 - High Leaf Wetness",
        "Soil/Leaf 3 - Low Soil Moisture",
        "Soil/Leaf 3 - High Soil Moisture",
        "Soil/Leaf 3 - Low Leaf Temperature",
        "Soil/Leaf 3 - High Leaf Temperature",
        "Soil/Leaf 3 - Low Soil Temperature",
        "Soil/Leaf 3 - High Soil Temperature",
    },
    {
        "Soil/Leaf 4 - Low Leaf Wetness",
        "Soil/Leaf 4 - High Leaf Wetness",
        "Soil/Leaf 4 - Low Soil Moisture",
        "Soil/Leaf 4 - High Soil Moisture",
        "Soil/Leaf 4 - Low Leaf Temperature",
        "Soil/Leaf 4 - High Leaf Temperature",
        "Soil/Leaf 4 - Low Soil Temperature",
        "Soil/Leaf 4 - High Soil Temperature",
    }
};


////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
LoopPacket::LoopPacket(void) : log(VP2Logger::getLogger("LoopPacket")) {
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
LoopPacket::~LoopPacket(void) {
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
int
LoopPacket::getNextRecord() const {
    return nextRecord;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
LoopPacket::BaroTrend
LoopPacket::getBaroTrend() const {
    return baroTrend;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Temperature> &
LoopPacket::getOutsideTemperature() const {
    return outsideTemperature;
}
        
////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Temperature> &
LoopPacket::getInsideTemperature() const {
    return insideTemperature;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Humidity> &
LoopPacket::getOutsideHumidity() const {
    return outsideHumidity;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Humidity> &
LoopPacket::getInsideHumidity() const {
    return insideHumidity;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Speed> &
LoopPacket::getWindSpeed() const {
    return windSpeed;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Speed> &
LoopPacket::getAvgWindSpeed10Min() const {
    return avgWindSpeed10Min;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Heading> &
LoopPacket::getWindDirection() const {
    return windDirection;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Pressure> &
LoopPacket::getBarometricPressure() const {
    return barometricPressure;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
Rainfall
LoopPacket::getRainRate() const {
    return rainRate;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
Rainfall
LoopPacket::getStormRain() const {
    return stormRain;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
DateTime
LoopPacket::getStormStart() const {
    return stormStart;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
Rainfall
LoopPacket::getDayRain() const {
    return dayRain;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
Rainfall
LoopPacket::getMonthRain() const {
    return monthRain;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
Rainfall
LoopPacket::getYearRain() const {
    return yearRain;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<UvIndex> &
LoopPacket::getUvIndex() const {
    return uvIndex;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Evapotranspiration> &
LoopPacket::getYearET() const {
    return yearET;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Evapotranspiration> &
LoopPacket::getMonthET() const {
    return monthET;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Evapotranspiration> &
LoopPacket::getDayET() const {
    return dayET;
}


////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<SolarRadiation> &
LoopPacket::getSolarRadiation() const {
    return solarRadiation;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
float
LoopPacket::getConsoleBatteryVoltage() const {
    return consoleBatteryVoltage;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
LoopPacket::Forecast
LoopPacket::getForecastIcon() const {
    return forecastIcon;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
int
LoopPacket::getForecastRule() const {
    return forecastRule;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool
LoopPacket::isTransmitterBatteryGood(int index) const {
    return (transmitterBatteryStatus & (1 << (index - 1))) == 0;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Temperature> &
LoopPacket::getExtraTemperature(int index) const {
    return extraTemperature[index];
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Humidity> &
LoopPacket::getExtraHumidity(int index) const {
    return extraHumidity[index];
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<LeafWetness> &
LoopPacket::getLeafWetness(int index) const {
    return leafWetness[index];
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<SoilMoisture> &
LoopPacket::getSoilMoisture(int index) const {
    return soilMoisture[index];
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Temperature> &
LoopPacket::getSoilTemperature(int index) const {
    return soilTemperature[index];
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const Measurement<Temperature> &
LoopPacket::getLeafTemperature(int index) const {
    return leafTemperature[index];
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool
LoopPacket::isStormOngoing() const {
    //
    // The last couple of LOOP packets that have a valid storm start will
    // report a storm rain total of 0.0 inches. This may be an indicator that the storm has stopped,
    // but we are not using that at this point in time. By definition the storm rain has to be > 0, so
    // we will stop reporting an ongoing storm if the storm rain is 0.0
    //
    return stormStart != 0 && stormRain > 0.0;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool
LoopPacket::decodeLoopPacket(byte buffer[]) {
    //
    // Perform a number of validation on the Loop packet before decoding all of the values
    //
    if (buffer[0] != 'L' || buffer[1] != 'O' || buffer[2] != 'O') {
        log.log(VP2Logger::VP2_ERROR) << "LOOP buffer does not begin with LOO: "
                                      << "[0] = " << buffer[0] << " [1] = " << buffer[1] << " [2] = " << buffer[2] << endl;
        return false;
    }

    if (!VantagePro2CRC::checkCRC(buffer, 97)) {
        log.log(VP2Logger::VP2_ERROR) << "LOOP packet failed CRC check" << endl;
        return false;
    }


    int packetType = BitConverter::toInt8(buffer, 4);
    if (packetType != LOOP_PACKET_TYPE) {
        log.log(VP2Logger::VP2_ERROR)<< "Invalid packet type for LOOP packet. Expected: "
                                     << LOOP_PACKET_TYPE << " Received: " << packetType << endl;
        return false;
    }

    if (buffer[95] != VP2Constants::LINE_FEED || buffer[96] != VP2Constants::CARRIAGE_RETURN) {
        log.log(VP2Logger::VP2_ERROR) << "<LF><CR> not found" << endl;
        return false;
    }


    if (buffer[3] != 'P') {
        switch (BitConverter::toInt8(buffer, 3)) {
            case 255:
                baroTrend = UNKNOWN;
                break;
            case 196:
                baroTrend = FALLING_RAPIDLY;
                break;
            case 236:
                baroTrend = FALLING_SLOWLY;
                break;
            case 0:
                baroTrend = STEADY;
                break;
            case 20:
                baroTrend = RISING_SLOWLY;
                break;
            case 60:
                baroTrend = RISING_RAPIDLY;
                break;
            default:
                log.log(VP2Logger::VP2_ERROR) << "Invalid barometer trend 0x" << hex << (int)buffer[3] << dec << endl;
                return false;
        }
    }

    nextRecord = BitConverter::toInt16(buffer,5);

    bool valid;
    VP2Decoder::decodeBarometricPressure(buffer, 7, barometricPressure);
    VP2Decoder::decode16BitTemperature(buffer, 9, insideTemperature);
    VP2Decoder::decodeHumidity(buffer, 11, insideHumidity);
    VP2Decoder::decode16BitTemperature(buffer, 12, outsideTemperature);

    windSpeed = VP2Decoder::decodeWindSpeed(buffer, 14);
    avgWindSpeed10Min = VP2Decoder::decodeWindSpeed(buffer, 15);
    windDirection = VP2Decoder::decodeWindDirection(buffer, 16);

    for (int i = 0; i < VP2Constants::MAX_EXTRA_TEMPERATURES; i++)
        VP2Decoder::decode8BitTemperature(buffer, 18 + i, extraTemperature[i]);

    for (int i = 0; i < VP2Constants::MAX_SOIL_TEMPERATURES; i++)
        VP2Decoder::decode8BitTemperature(buffer, 25 + i, soilTemperature[i]);

    for (int i = 0; i < VP2Constants::MAX_LEAF_TEMPERATURES; i++)
        VP2Decoder::decode8BitTemperature(buffer, 29 + i, leafTemperature[i]);

    VP2Decoder::decodeHumidity(buffer, 33, outsideHumidity);

    for (int i = 0; i < VP2Constants::MAX_EXTRA_HUMIDITIES; i++)
        VP2Decoder::decodeHumidity(buffer, 34 + i, extraHumidity[i]);

    rainRate = VP2Decoder::decodeRain(buffer, 41);

    VP2Decoder::decodeUvIndex(buffer, 43, uvIndex);
    VP2Decoder::decodeSolarRadiation(buffer, 44, solarRadiation);

    stormRain = VP2Decoder::decodeStormRain(buffer, 46);
    stormStart = VP2Decoder::decodeStormStartDate(buffer, 48);

    dayRain = VP2Decoder::decodeRain(buffer, 50);
    monthRain = VP2Decoder::decodeRain(buffer, 52);
    yearRain = VP2Decoder::decodeRain(buffer, 54);

    dayET = VP2Decoder::decodeDayET(buffer, 56);
    monthET = VP2Decoder::decodeMonthYearET(buffer, 58);
    yearET = VP2Decoder::decodeMonthYearET(buffer, 60);

    for (int i = 0; i < VP2Constants::MAX_SOIL_MOISTURES; i++)
        VP2Decoder::decodeSoilMoisture(buffer, 62 + i, soilMoisture[i]);

    for (int i = 0; i < VP2Constants::MAX_LEAF_WETNESSES; i++)
        VP2Decoder::decodeLeafWetness(buffer, 66 + i, leafWetness[i]);

    for (int i = 0; i < 16; i++) {
        int alarms = BitConverter::toInt8(buffer, 70 + i);
        for (int j = 0; j < 8; j++) {
            if (alarms & (1 << j)) {
                std::cout << lookupAlarm(i, j) << std::endl;
            }
        }
    }

    transmitterBatteryStatus = BitConverter::toInt8(buffer, 86);
    log.log(VP2Logger::VP2_DEBUG2) << "Transmitter Battery Status: " << transmitterBatteryStatus << endl;

    consoleBatteryVoltage = VP2Decoder::decodeConsoleBatteryVoltage(buffer, 87);
    log.log(VP2Logger::VP2_DEBUG2) << "Console Battery Voltage: " << consoleBatteryVoltage << endl;

    forecastIcon = static_cast<Forecast>(BitConverter::toInt8(buffer, 89));
    forecastRule = BitConverter::toInt8(buffer, 90);

    sunriseTime = VP2Decoder::decodeTime(buffer, 91);
    sunsetTime = VP2Decoder::decodeTime(buffer, 93);

    return true;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
string
LoopPacket::lookupAlarm(int byte, int bit) const {
    if (byte < 0 || byte >= 16 || bit < 0 || bit >= 8)
        return "Invalid Alarm Index";

    const char * alarmString = ALARM_STRINGS[byte][bit];

    if (alarmString == nullptr)
        return "Invalid Alarm Index";
    else
        return alarmString;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
string
LoopPacket::getBaroTrendString() const {
    switch(baroTrend) {
        case FALLING_RAPIDLY:
            return "FALLING_RAPIDLY";

        case FALLING_SLOWLY:
            return "FALLING_SLOWLY";

        case STEADY:
            return "STEADY";

        case RISING_SLOWLY:
            return "RISING_SLOWLY";

        case RISING_RAPIDLY:
            return "RISING_RAPIDLY";

        default:
            return "STEADY";
    }
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
string
LoopPacket::getForecastIconString() const {
    switch (forecastIcon) {
        case SUNNY:
            return "SUNNY";

        case PARTLY_CLOUDY:
            return "PARTLY_CLOUDY";

        case MOSTLY_CLOUDY:
            return "MOSTLY_CLOUDY";

        case MOSTLY_CLOUDY_WITH_RAIN:
            return "MOSTLY_CLOUDY_WITH_RAIN";

        case MOSTLY_CLOUDY_WITH_SNOW:
            return "MOSTLY_CLOUDY_WITH_SNOW";

        case MOSTLY_CLOUDY_WITH_RAIN_OR_SNOW:
            return "MOSTLY_CLOUDY_WITH_RAIN_OR_SNOW";

        case PARTLY_CLOUDY_WITH_RAIN_LATER:
            return "PARTLY_CLOUDY_WITH_RAIN_LATER";

        case PARTLY_CLOUDY_WITH_SNOW_LATER:
            return "PARTLY_CLOUDY_WITH_SNOW_LATER";

        case PARTLY_CLOUDY_WITH_RAIN_OR_SNOW_LATER:
            return "PARTLY_CLOUDY_WITH_SNOW_LATER";

        default:
            return "SUNNY";
    }
}

}
